divintseg/_core.py

```diff
@@ -422,3 +422,151 @@ def similarity(
+    follows:
+
+    +----------+-------------+-----------------------+
+    | Region   | Subregion   | Likelihood of an S    |
+    +==========+=============+=======================+
+    | Region 1 | Subregion A | 100 / (100 + 0) = 1.0 |
+    +----------+-------------+-----------------------+
+    | Region 1 | Subregion B | 50 / (50 + 50) = 0.5  |
+    +----------+-------------+-----------------------+
+    | Region 2 | Subregion C | 0 / (0 + 100) = 0.0   |
+    +----------+-------------+-----------------------+
+    | Region 2 | Subregion D | 0 / (0 + 50) = 0.0    |
+    +----------+-------------+-----------------------+
+    | Region 2 | Subregion E | 10 / (10 + 90) = 0.1  |
+    +----------+-------------+-----------------------+
+
+    Next, we can compute the fraction of all S's in
+    each subregion of each region. There are 150 S's
+    in Region 1 and 10 S's in region 2, therefore, we have:
+
+    +----------+-------------+--------------------------------+
+    | Region   | Subregion   | Fraction of all As in Region   |
+    +==========+=============+================================+
+    | Region 1 | Subregion A | 100 / 150 = 0.6667             |
+    +----------+-------------+--------------------------------+
+    | Region 1 | Subregion B | 50 / 150 = 0.3333              |
+    +----------+-------------+--------------------------------+
+    | Region 2 | Subregion C | 0 / 10 = 0.0000                |
+    +----------+-------------+--------------------------------+
+    | Region 2 | Subregion D | 0 / 10 = 0.0000                |
+    +----------+-------------+--------------------------------+
+    | Region 2 | Subregion E | 10 / 10 = 1.0000               |
+    +----------+-------------+--------------------------------+
+
+    Finally, for each subregion, we multiply these together and
+    add them up the values for the subregions in each region.
+    For Region 1, we get
+
+    .. math::
+        (0.6667 * 1.0) + (0.3333 * 0.5) = 0.8333.
+
+    For Region 2, we get
+
+    .. math::
+        0.0 * 0.0 + 0.0 * 0.0 + 1.000 * 0.1 = 0.1.
+
+    Note that the implentation may not do this math exactly as
+    specified here, but it will do something equivalent.
+    """
+    if by is None:
+        df_communities["REGION"] = "Region " + df_communities.index.astype(str)
+        by = "REGION"
+    if over is None:
+        df_communities["SUBREGION"] = "Subregion " + df_communities.index.astype(str)
+        over = "SUBREGION"
+    else:
+        subregion_groups = df_communities.groupby(over)
+        for _, group in subregion_groups:
+            if len(group[by].unique()) > 1:
+                raise ValueError("Subregions must have the same region")
+                # If 2 subregions are the same, their region must be the same
+    df_by = df_communities.groupby(by).sum(numeric_only=True)[group_name]
+    df_over = df_communities.groupby(over).sum(numeric_only=True)
+    likelihood = df_over[group_name] / df_over.sum(axis="columns")
```

Comment:
I don't know if this would help make your code a bit simpler, but you can group by multiple columns at once. For example, `df_communities.groupby(['by', 'over']).sum(numeric_only=True)` will sum things up for each subregion in each region. It also handles the case discussed above where the same subregion identifier occurs in two different regions. I think you can probably refactor your code a bit to use this multi-column style of `groupby` to make your code a little simpler.